source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -17,13 +21,35 @@ void ProcessorState::startMessageTimer(Event::TimerCb cb, std::chrono::milliseco
   message_timer_->enableTimer(timeout);
 }
 
-bool ProcessorState::handleHeadersResponse(const HeadersResponse& response) {
+bool ProcessorState::handleHeadersResponse(const HeadersResponse& response,
+                                           ProcessingMode_BodySendMode body_mode) {
   if (callback_state_ == CallbackState::Headers) {
     ENVOY_LOG(debug, "applying headers response");
     MutationUtils::applyCommonHeaderResponse(response, *headers_);
-    headers_ = nullptr;
     callback_state_ = CallbackState::Idle;
+    clearWatermark();
     message_timer_->disableTimer();
+
+    if (body_mode == ProcessingMode::BUFFERED) {
+      if (body_send_deferred_) {
+        // If we get here, then all the body data came in before the header message
+        // was complete, and the server wants the body. So, don't continue filter
+        // processing, but send the buffered request body now.
+        ENVOY_LOG(debug, "Sending buffered request body message");
+        callback_state_ = CallbackState::BufferedBody;
+        startMessageTimer(std::bind(&Filter::onMessageTimeout, &filter_),
+                          filter_.config().messageTimeout());
+        filter_.sendBufferedData(*this, true);
+      }
+
+      // Otherwise, we're not ready to continue processing because then
+      // we won't be able to modify the headers any more, so do nothing and
+      // let the doData callback handle body chunks until the end is reached.
+      return true;
```

Comment:
Hmm, shouldn't this be false to indicate that we did not process the message? I don't recall how this was in the previous iteration, but now it seems like we return true regardless of what `body_send_deferred_` is set to?

Should there be a return after L42?

Reply:
The contract for this function, which is called from "onReceiveMessage" in ext_proc.cc, is that it returns false if the message is unexpected (aka spurious). So it's correct to return true in this case, and that's how it's worked for a long time. 

In that particular case -- line 48 -- a "headers response" came back from the server in response to a previous headers request. The processor requested to buffer the body, but the entire body hadn't been received by the filter yet, which is why "body_send_deferred_" is false. The filter needs to continue buffering the body until the end of stream is reached, at which point it will send the body to the server. So yes, that's correct in this case.

Reply:
Got it! I saw 
```
    message_handled = decoding_state_.handleHeadersResponse(response->request_headers());
```
and it made me think that we were going to return false in this case since we hadn't been able to "handle" the response (due to the body not being fully buffered). Your explanation makes sense though, thanks for that!